File: NovilhaNutri/Core/Modelos/Insumo.swift
import Foundation

/// Forma de aquisição do insumo, usada na conversão de quilos para sacas.
struct Embalagem: Codable, Hashable {
    enum Tipo: String, Codable, CaseIterable, Identifiable, Hashable {
        case saca
        case granel
        case pastejo

        var id: String { rawValue }

        var nome: String {
            switch self {
            case .saca: return "Saca"
            case .granel: return "Granel (tonelada)"
            case .pastejo: return "Pastejo (não comprado)"
            }
        }
    }

    var tipo: Tipo
    /// Peso líquido da saca em quilos. Relevante apenas quando `tipo == .saca`.
    var kgPorSaca: Double

    init(tipo: Tipo = .saca, kgPorSaca: Double = 50) {
        self.tipo = tipo
        self.kgPorSaca = kgPorSaca
    }

    /// Tamanhos de saca praticados no mercado brasileiro de insumos.
    static let tamanhosPadrao: [Double] = [60, 50, 40, 30, 25, 20]

    static let saca60 = Embalagem(tipo: .saca, kgPorSaca: 60)
    static let saca50 = Embalagem(tipo: .saca, kgPorSaca: 50)
    static let saca40 = Embalagem(tipo: .saca, kgPorSaca: 40)
    static let saca30 = Embalagem(tipo: .saca, kgPorSaca: 30)
    static let saca25 = Embalagem(tipo: .saca, kgPorSaca: 25)
    static let granel = Embalagem(tipo: .granel, kgPorSaca: 1000)
    static let pastejo = Embalagem(tipo: .pastejo, kgPorSaca: 0)

    var descricao: String {
        switch tipo {
        case .saca: return "Saca de \(Formatadores.numero(kgPorSaca, casas: 0)) kg"
        case .granel: return "Granel (tonelada)"
        case .pastejo: return "Pastejo"
        }
    }

    /// Quantidade de quilos que corresponde a uma unidade de compra.
    /// Retorna `nil` quando o insumo não é adquirido (pastejo).
    var kgPorUnidade: Double? {
        switch tipo {
        case .saca: return kgPorSaca > 0 ? kgPorSaca : nil
        case .granel: return 1000
        case .pastejo: return nil
        }
    }

    var nomeUnidade: String {
        switch tipo {
        case .saca: return "saca"
        case .granel: return "tonelada"
        case .pastejo: return "-"
        }
    }

    var nomeUnidadePlural: String {
        switch tipo {
        case .saca: return "sacas"
        case .granel: return "toneladas"
        case .pastejo: return "-"
        }
    }
}

/// Alimento disponível na propriedade, com sua composição bromatológica.
/// Teores de PB e NDT são sempre expressos em percentual da matéria seca.
struct Insumo: Identifiable, Codable, Hashable {
    var id: UUID
    var nome: String
    var categoria: CategoriaInsumo
    /// Matéria seca (% da matéria natural).
    var materiaSeca: Double
    /// Proteína bruta (% da matéria seca).
    var proteinaBruta: Double
    /// Nutrientes digestíveis totais (% da matéria seca).
    var ndt: Double
    var embalagem: Embalagem
    /// Preço por unidade de compra (por saca, ou por tonelada no granel).
    var precoUnitario: Double
    var observacao: String

    init(id: UUID = UUID(),
         nome: String,
         categoria: CategoriaInsumo,
         materiaSeca: Double,
         proteinaBruta: Double,
         ndt: Double,
         embalagem: Embalagem = .saca50,
         precoUnitario: Double = 0,
         observacao: String = "") {
        self.id = id
        self.nome = nome
        self.categoria = categoria
        self.materiaSeca = materiaSeca
        self.proteinaBruta = proteinaBruta
        self.ndt = ndt
        self.embalagem = embalagem
        self.precoUnitario = precoUnitario
        self.observacao = observacao
    }

    /// Fração de matéria seca (0 a 1), protegida contra valores inválidos.
    var fracaoMateriaSeca: Double {
        min(max(materiaSeca / 100, 0.01), 1.0)
    }

    /// Converte quilos de matéria seca em quilos de matéria natural (como fornecido).
    func materiaNatural(deMateriaSeca kgMS: Double) -> Double {
        kgMS / fracaoMateriaSeca
    }

    /// Converte quilos de matéria natural em quilos de matéria seca.
    func materiaSeca(deMateriaNatural kgMN: Double) -> Double {
        kgMN * fracaoMateriaSeca
    }

    /// Preço por quilo de matéria natural.
    var precoPorKg: Double {
        guard let kg = embalagem.kgPorUnidade, kg > 0 else { return 0 }
        return precoUnitario / kg
    }

    /// Preço por quilo de matéria seca, útil para comparar alimentos.
    var precoPorKgMateriaSeca: Double {
        precoPorKg / fracaoMateriaSeca
    }

    var resumoBromatologico: String {
        "MS \(Formatadores.percentual(materiaSeca)) - PB \(Formatadores.percentual(proteinaBruta)) - NDT \(Formatadores.percentual(ndt))"
    }
}

File: web/src/nucleo/insumo.ts
/**
 * Alimento disponível na propriedade e a forma como ele é comprado.
 *
 * Porte de `Core/Modelos/Insumo.swift`, do antigo aplicativo de iPhone.
 * Teores de PB e NDT são sempre
 * percentuais da matéria seca.
 */
import type { CategoriaInsumo } from "./classificacoes.js";
import { numero, percentual } from "./formatadores.js";

// ---------------------------------------------------------------- embalagem

export type TipoEmbalagem = "saca" | "granel" | "pastejo";

export interface Embalagem {
  tipo: TipoEmbalagem;
  /** Peso líquido da saca em quilos. Só importa quando o tipo é "saca". */
  kgPorSaca: number;
}

/** Tamanhos de saca praticados no mercado brasileiro de insumos. */
export const TAMANHOS_PADRAO: readonly number[] = [60, 50, 40, 30, 25, 20];

export function embalagem(tipo: TipoEmbalagem = "saca", kgPorSaca = 50): Embalagem {
  return { tipo, kgPorSaca };
}

export const SACA_60 = embalagem("saca", 60);
export const SACA_50 = embalagem("saca", 50);
export const SACA_40 = embalagem("saca", 40);
export const SACA_30 = embalagem("saca", 30);
export const SACA_25 = embalagem("saca", 25);
export const GRANEL = embalagem("granel", 1000);
export const PASTEJO = embalagem("pastejo", 0);

/**
 * Quilos que correspondem a uma unidade de compra.
 * Devolve `null` quando o insumo não é adquirido (pastejo).
 */
export function kgPorUnidade(e: Embalagem): number | null {
  switch (e.tipo) {
    case "saca":
      return e.kgPorSaca > 0 ? e.kgPorSaca : null;
    case "granel":
      return 1000;
    case "pastejo":
      return null;
  }
}

export function nomeUnidade(e: Embalagem): string {
  switch (e.tipo) {
    case "saca":
      return "saca";
    case "granel":
      return "tonelada";
    case "pastejo":
      return "-";
  }
}

export function nomeUnidadePlural(e: Embalagem): string {
  switch (e.tipo) {
    case "saca":
      return "sacas";
    case "granel":
      return "toneladas";
    case "pastejo":
      return "-";
  }
}

export function descricaoEmbalagem(e: Embalagem): string {
  switch (e.tipo) {
    case "saca":
      return `Saca de ${numero(e.kgPorSaca, 0)} kg`;
    case "granel":
      return "Granel (tonelada)";
    case "pastejo":
      return "Pastejo";
  }
}

// ------------------------------------------------------------------ insumo

export interface Insumo {
  id: string;
  nome: string;
  categoria: CategoriaInsumo;
  /** Matéria seca (% da matéria natural). */
  materiaSeca: number;
  /** Proteína bruta (% da matéria seca). */
  proteinaBruta: number;
  /** Nutrientes digestíveis totais (% da matéria seca). */
  ndt: number;
  embalagem: Embalagem;
  /** Preço por unidade de compra: por saca, ou por tonelada no granel. */
  precoUnitario: number;
  observacao: string;
}

export function criarInsumo(
  entrada: Pick<Insumo, "nome" | "categoria" | "materiaSeca" | "proteinaBruta" | "ndt"> &
    Partial<Insumo>,
): Insumo {
  return {
    id: crypto.randomUUID(),
    embalagem: SACA_50,
    precoUnitario: 0,
    observacao: "",
    ...entrada,
  };
}

/** Fração de matéria seca (0 a 1), protegida contra valores inválidos. */
export function fracaoMateriaSeca(insumo: Insumo): number {
  return Math.min(Math.max(insumo.materiaSeca / 100, 0.01), 1.0);
}

/** Converte quilos de matéria seca em quilos de matéria natural. */
export function materiaNatural(insumo: Insumo, kgMS: number): number {
  return kgMS / fracaoMateriaSeca(insumo);
}

/** Converte quilos de matéria natural em quilos de matéria seca. */
export function materiaSecaDe(insumo: Insumo, kgMN: number): number {
  return kgMN * fracaoMateriaSeca(insumo);
}

/** Preço por quilo de matéria natural. */
export function precoPorKg(insumo: Insumo): number {
  const kg = kgPorUnidade(insumo.embalagem);
  if (kg === null || kg <= 0) return 0;
  return insumo.precoUnitario / kg;
}

/** Preço por quilo de matéria seca, útil para comparar alimentos. */
export function precoPorKgMateriaSeca(insumo: Insumo): number {
  return precoPorKg(insumo) / fracaoMateriaSeca(insumo);
}

export function resumoBromatologico(insumo: Insumo): string {
  return (
    `MS ${percentual(insumo.materiaSeca)} - ` +
    `PB ${percentual(insumo.proteinaBruta)} - ` +
    `NDT ${percentual(insumo.ndt)}`
  );
}
